### ISSUE TITLE
Parallelized computing

### ISSUE THREAD

**Issue by Bai-YT** [fra31/auto-attack#94]
Hi,

I am trying to run `AutoAttack` on multiple GPUs to accelerate the experiments. However, it seems that simply wrapping the model with `torch.nn.DataParallel` will result in one GPU running at nearly 100% with other GPUs only at about 20%. So I just hoped to ask if there is a more efficient way of parallelization. Has there been a `DistributedDataParallel` code example for `AutoAttack`? Thank you very much.

**Comment by fra31** [comment]
Hi,

I haven't tried myself, but if I remember right @max-andr used AA with parallelization at some point, maybe he has some insights.

**Comment by max-andr** [comment]
Hi,

actually, I haven't tried it myself, so can't really tell, sorry :/

**Comment by Bai-YT** [comment]
Thank you @max-andr @fra31 for the information!

**Comment by CNOCycle** [comment]
In my forked repo, `torch.nn.DataParallel` works properly. When the evaluation is run on two GPUs, it reduces the elapsed time by about half. Please share your code to further investigation.

Running auto-attack with `DistributedDataParallel` is much complex. Some control flows should be refined to retrieve intermediate information efficiently or avoid deadlocks. Those mechanisms or programming logic highly depends how you launch DDP version of auto-attack. However, if all you need is the final robust accuracy, about 20 lines modifications are required.

**Comment by Bai-YT** [comment]
> In my forked repo, `torch.nn.DataParallel` works properly. When the evaluation is run on two GPUs, it reduces the elapsed time by about half. Please share your code to further investigation.
> 
> Running auto-attack with `DistributedDataParallel` is much complex. Some control flows should be refined to retrieve intermediate information efficiently or avoid deadlocks. Those mechanisms or programming logic highly depends how you launch DDP version of auto-attack. However, if all you need is the final robust accuracy, about 20 lines modifications are required.

Thank you for sharing your experience! Much appreciated.

**Comment by MaxHeuillet** [comment]
Hello, I am facing the same issue when evaluating with a model wrapped with DDP. I am having a hard time finding a solution. The goal is simply to evaluate a model and leverage multiple GPUs available, rather than 1. The bug persists even after I remove the DDP wrapper.  Did anyone find of a fix? Best,

**Comment by legolasflagstaff** [comment]
Hello, a related issue can be found here ([https://github.com/fra31/auto-attack/issues/84](https://github.com/fra31/auto-attack/issues/84)). AutoAttack does not work well when the distributed backend is configured with NCCL. 

- Instead of configuring your distributed testing setup with NCCL (e.g. ```dist.init_process_group("nccl", rank=rank, world_size=world_size)``` ), launch independent processes for each rank:

```python
import torch.multiprocessing as mp 
# Launch distributed testing processes
processes = []
for rank in range(self.world_size):
     p = mp.Process(target=self.test, args=(rank, ) )
     p.start()
     processes.append(p)

# Wait for all processes to finish
for p in processes:
     p.join()

```
- When creating the ```AutoAttack``` adversary instance in the ```test``` function, make sure that you pass the ```rank``` argument in ```AutoAttack```.

- If you used DDP wrapper and NCCL for training before testing, don't forget to close the nccl connection ```dist.destroy_process_group()``` before testing. Also, save the trained state dict appropriately to remove all the DDP components. Make sure your model is back on cpu when you save the state_dict.
```python
if rank == 0:
     model_to_save = model.module
     model_to_save = model_to_save.cpu()  
     torch.save(model_to_save.state_dict(), './state_dicts/trained_model.pt')
```

- Then, in the test function, load back the state dict with  
```python
trained_state_dict = torch.load('./state_dicts/trained_model.pt', map_location='cpu')
model.load_state_dict(trained_state_dict)
model.to(rank)
```
